### ISSUE TITLE
Unstable refresh rate when VRR and HDR are enabled

### ISSUE THREAD

**Issue by mkopec** [ValveSoftware/gamescope#1006]
Problem:

When running games via steam in gamescope in DRM mode with HDR and VRR enabled, the screen's refresh rate is very unstable and jumps between 61 and 120 instead of being consistent with the game's framerate. This results in visible stuttering instead of smooth gameplay.

My setup is:
- CPU: Ryzen 5700G
- GPU: Radeon 6700XT
- Kernel: 6.5.9-arch2 (although amd-drm-next exhibits the same problem) 
- Display: Samsung S95B TV, 3840x2160@120Hz (with chroma subsampling since AMD doesn't yet support HDMI 2.1, no idea if this is relevant at all)

Games tested:
- The Talos Principle 2
- Forza Horizon 4
- Doom Eternal

All of these run below the screen's native 120Hz refresh rate, mostly around 80-90FPS. I'm checking the TV's refresh rate via its builtin OSD, it's telling me that Freesync is enabled and the refresh rate is rapidly switching between 120 and 61. If a game is running at 120FPS stably, then the issue does not occur.

Interestingly, in Steam UI I don't see this problem, instead I see a stable 110 FPS and no stutter at all.

**Comment by lextra2** [comment]
Are you playing with a controller? I've heard mouse inputs can cause issues with vrr, so maybe unplug keyboard and mouse and try playing with only a controller, to see if that makes a difference.

**Comment by mkopec** [comment]
Hmm, yes, I'm playing with an Xbox Series X/S controller via the official dongle, but the same issue appears when I use a mouse polling at 1000Hz.

**Comment by Joshua-Ashton** [comment]
Is this problem present on other VRR implementations for you? I can't see anything that'd cause this on our side and it definitely doesn't happen on my display. Likely something kernel sided?

**Comment by mkopec** [comment]
VRR works well in Sway for me with the same games.

Did some more testing, started removing options from my gamescope command until the problem disappeared. It looks like `--hdr-enabled` is the culprit, if enabled I see this weird behavior, if disabled, VRR works perfectly again.

Can't really test HDR + VRR in any other compositors 😅 Do you think this might be a kernel bug @Joshua-Ashton ?

**Comment by Lawstorant** [comment]
Same thing happens on my 5950X + 6800XT PC while running gamescope i DRM mode. `--hdr-enabled` breaks VRR and it jumps between 120 Hz and 61 Hz. Sometimes it just gets stuck at 66 Hz when using steam with `-steamdeck -steamos3` arguments

**Comment by mkopec** [comment]
Issue appears to be fixed as of commit d434d86c0d8a569af518c6bdd9ff2616f5a784c9 , Linux 6.6.7-arch1-1, Mesa 23.3.1. Thanks!

**Comment by Lawstorant** [comment]
I did some further testing and HDR + VRR is still broken, even with 6.7 kernel. What's weird is that VRR + HDR works with nested gamescope needed to enable HDR gaming on plasma 6 (+ wsi layer).

**Comment by mkopec** [comment]
Indeed the issue seems to still be there as of the latest revision and kernel 6.7.0. I only tested Doom Eternal when I reported that the issue is fixed, but after some further testing, other games still have the same issue, and with the latest revisions, so does Doom Eternal again.

**Comment by Lawstorant** [comment]
@Joshua-Ashton as there are DiplayPort->HDMI 2.1 PCONs that are whitelisted to support VRR in the amdgpu, HDR + VRR in gamescpoe should become a big priority, as this is (or will be) supported by the SteamDeck. The VRR allowlist was backported to 6.1 LTS kernel.

**Comment by Joshua-Ashton** [comment]
We don't composite in the Steam Deck as we have everything for HDR multiplane scanout in SteamOS.

If composite is taking 4+ ms, then something is horribly wrong.

**Comment by Lawstorant** [comment]
Is it possible that you're doing something in SteamOS that wouldn't be in stock ArchLinux? It's the same amdgpu + gamescope in DRM mode after all

**Comment by Lawstorant** [comment]
@mkopec thanks for confirming that. Sadly, only 6.7 fixes HDR colorspace on DisplayPort with amdgpu, so I think I'll have to apply these patches manually on top of 6.7

**Comment by ajoneil** [comment]
I can confirm I'm seeing the same issue on stock Arch with an RX 6800XT, and using the neptune kernel fixes it for me. This is with a Cable Matters 102101 DP -> HDMI 2.1 adapter, with an LG B9 with a patched EDID to add the missing freesync range.

**Comment by dklementowski** [comment]
> I can confirm I'm seeing the same issue on stock Arch with an RX 6800XT, and using the neptune kernel fixes it for me. This is with a Cable Matters 102101 DP -> HDMI 2.1 adapter, with an LG B9 with a patched EDID to add the missing freesync range.

How did you manage to run neptune kernel with RX 6800XT? I tried linux-neptune-65 from AUR, but I had to edit config so that it included BTRFS explicitly (otherwise mounting btrfs volumes on boot would fail), but even then amdgpu isn't loaded, there's nothing in dmesg about it and Plasma 6 uses llvmpipe as its renderer (gamescope session obviously wouldn't work).

I've got the same Cable Matters adapter with flashed firmware (that was linked in the well known Freedesktop's Gitlab issue about HDMI 2.1 support) and I managed to get it working using linux-amd-color and gamescope-amd-color (which as I understand, includes your patches from https://github.com/ValveSoftware/gamescope/pull/1149), but it's  unstable with Gamescope, and Plasma 6 has messed up colors with HDR turned on (on a contrary, Plasma 6 worked nicely with HDR enabled and VRR on stock kernel), and sometimes the system freezes completely. When I just launch a game and keep playing it without changing any settings/showing Steam overlay, it works really well.

**Comment by WeslyG** [comment]
I am using a philips 55oled807 with an hdmi 2.1 cord, and an i5 12400f processor, an amd RX 7800 XT graphics card with gnome shell os, and when I switch my TV to hdr + vrr mode, I see flickering images, when I wake up from sleep I can get a black screen while I turn off my TV to hdr only mode (without vrr). Then everything will work stably. I installed the latest OS and updated everything that was updated) hdr+ vrr does not work stably for me, even with the latest hardware unfortunately.

**Comment by dklementowski** [comment]
> I am using a philips 55oled807 with an hdmi 2.1 cord, and an i5 12400f processor, an amd RX 7800 XT graphics card with gnome shell os, and when I switch my TV to hdr + vrr mode, I see flickering images, when I wake up from sleep I can get a black screen while I turn off my TV to hdr only mode (without vrr). Then everything will work stably. I installed the latest OS and updated everything that was updated) hdr+ vrr does not work stably for me, even with the latest hardware unfortunately.

This issue is about running Gamescope in DRM mode, aka SteamDeck mode, which has support for both features on SteamOS, but it malfunctions on mainline kernel. GNOME doesn't support VRR or HDR yet (though VRR looks like is close)

**Comment by Lawstorant** [comment]
@mkopec probably goo to close this one now

**Comment by maximilienleclei** [comment]
@Lawstorant is it fixed?

**Comment by Lawstorant** [comment]
Yes, has been for a long while now. I'm now basically always using gamescope session and it's near flawless. I think the only issue I still have is that there isn't a global setting for VRR to be only active during gaming

**Comment by mkopec** [comment]
Thanks, this has been resolved.
